File: src/model/bolt.ts
import {DTO} from "./dto";

export class BoltObj extends DTO {
    private readonly _id: string;
    private _contentID: string;
    private _head: string;
    private _length: number;
    private _threadSize: string;
    private _threadPitch: string;
    private _material: string;

    public static Parse(d: string): BoltObj {
        if (!d) {
            return BoltObj.Empty()
        }

        return new BoltObj(d);
    }

    public static Empty(): BoltObj {
        return new BoltObj({})
    }

    JSON(): object {
        console.log("bolt: ", this)
        return {
            id: this._id,
            content_id: this._contentID,
            head: this._head,
            length: this._length,
            thread_size: this._threadSize,
            thread_pitch: this._threadPitch,
            material: this._material
        }
    }

    private constructor(d: any) {
        super()
        
        this._id = d.id;
        this._contentID = d.content_id;
        this._head = d.head;
        this._length = d.length;
        this._threadSize = d.thread_size;
        this._threadPitch = d.thread_pitch;
        this._material = d.material;
    }

    //getters
    get id() {
        return this._id
    }

    get contentID() {
        return this._contentID
    }

    set contentID(value: string) {
        this._contentID = value
    }

    get length(): number {
        return this._length
    }

    set length(value: number) {
        if (value < 1) {
            throw new Error("length cannot be zero or negitive")
        }

        this._length = value
    }


    get head(): string {
        return this._head
    }

    set head(value: string) {
        this._head = value
    }

    get material(): string {
        return this._material
    }

    set material(value: string) {
        this._material = value
    }

    get threadSize(): string {
        return this._threadSize
    }

    set threadSize(value: string) {
        this._threadSize = value
    }

    get threadPitch(): string {
        return this._threadPitch
    }

    set threadPitch(value: string) {
        this._threadPitch = value
    }
}

File: app/model/bolt.go
package model

import (
	"time"

	"github.com/blevesearch/bleve/v2"
	"github.com/blevesearch/bleve/v2/mapping"
)

var _ IContent = (*Bolt)(nil)

type Bolt struct {
	//nolint:structcheck,unused
	tableName struct{} `pg:"goober.bolt,alias:bolt"`

	ID          string         `json:"id" pg:"id,pk,type:uuid"`
	ContentID   string         `json:"content_id" pg:"content_id,fk,type:uuid"`
	CreatedAt   time.Time      `json:"created_at" pg:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at" pg:"updated_at"`
	DeletedAt   *time.Time     `json:"deleted_at" pg:"deleted_at,soft_delete"`
	Head        BoltHeadType   `json:"head" pg:"head,type:goober.bolt_head_t"`
	Length      float32        `json:"length" pg:"length"`
	ThreadSize  string         `json:"thread_size" pg:"thread_size"`
	ThreadPitch string         `json:"thread_pitch" pg:"thread_pitch"`
	Material    MaterialFinish `json:"material" pg:"material,type:goober.material_finish_t"`
	Description string         `json:"description" pg:"description"`
}

func (b Bolt) String() string {
	return ""
}

func (b Bolt) BleveType() string {
	return "bolt"
}

func (b Bolt) CreateMapping(fieldMapping *mapping.FieldMapping) (string, *mapping.DocumentMapping) {
	boltMapping := bleve.NewDocumentMapping()

	boltMapping.AddFieldMappingsAt("head", fieldMapping)
	boltMapping.AddFieldMappingsAt("length", fieldMapping)
	boltMapping.AddFieldMappingsAt("thread_size", fieldMapping)
	boltMapping.AddFieldMappingsAt("thread_pitch", fieldMapping)
	boltMapping.AddFieldMappingsAt("material", fieldMapping)
	boltMapping.AddFieldMappingsAt("description", fieldMapping)

	return "bolt", boltMapping
}
